### DISCUSSION TITLE
Open3E - Adapter für ioBroker

### DISCUSSION THREAD

**Discussion by MyHomeMyData** [open3e/open3e#57] [Show and tell]
# Adapter für ioBroker führt "collect" und UDSonCAN zusammen

**Update an 30.01.2024:** Der Adapter ist jetzt im Beta-Repository von ioBroker verfügbar. Damit steht der normale Installation- und Update-Pfad zur Verfügung. Evtl. muss vorher auf das Beta-Repo umgestellt werden. Ich habe die Installationsanleitung entpsrechend aktualisiert, s.u.

In den letzten Wochen habe ich an einem Adapter für ioBroker gearbeitet, jetzt ist eine Testversion verfügbar:
* Lauschen auf dem internen und externen Bus (incl. Energy Meter E380)
* Lesen und Dekodieren aller Datenpunkte der gemeinsamen Datenpunkliste. Device spezifische Datenpunkte werden per RawCodec verarbeitet.
* **Schreiben aller für Datenpunkte**, die auf einer editierbaren White-List stehen. Auch komplexe Datenpunkte können geschrieben werden, z.B. kann nur die Anfangszeit für die Zirkulationspumpe eines bestimmten Zeitraumes geändert werden.
* Scan der Devices bei der ersten Einrichtung des Adapters
* Optional Festlegen des Gerätenamens für den Objektbaum
* Datenpunkt-Scan aller erkannten Devices (entspricht Depict-Funktion von open3E)
* Die Daten aus beiden Quellen (collect und UDS) werden pro Device in **einem** Objektbaum gesammelt.
* Für jeden Datenpunkt stehen die Format json, raw und tree parallel zur Verfügung.

**Ich würde mich freuen, wenn ihr den Adapter ausprobiert und mir Feedback gebt.** Da es meiner erster Adapter für ioBroker ist, kann ich nicht so richtig einschätzen, wie stabil das auf anderen System läuft. Also bitte immer mal in die Logs schauen. Da sich in der weiteren Entwicklung noch Dinge ändern können, bitte **den Adapter im Moment ausschließlich zum Testen nutzen**!

Die Installation muss im Moment noch per GitHub-Adresse gemacht werden. Wenn ein stabile Stand erreicht ist, werde ich den Adapter über den normalen ioBroker-Installationspfad verfügbar machen.

Den Adapter incl. README findet ihr [hier](https://github.com/MyHomeMyData/ioBroker.e3oncan).

**Comment by MyHomeMyData** [comment]
## Hier ein paar Screenshots des Adapters.

## Geräteliste
![e3oncan_devlist](https://github.com/open3e/open3e/assets/144950531/d86b6365-962a-449e-beed-8b719e70d93c)

## Konfiguration der gewünschten Daten
![e3oncan_extcan_scheds](https://github.com/open3e/open3e/assets/144950531/2a6bb97d-f3f1-4557-b629-4ee447552c73)

## Objektbaum
![e3oncan_object_tree](https://github.com/open3e/open3e/assets/144950531/35bfc310-0e03-4b8f-8b0e-b47c5b971f2b)

## Kleine Demo: Zeiträume für Zirkulationspumpe ändern
Ich habe im Adapter den json Datenpunkt 727 der Vitocal (DomesticHotWaterCirculationTimeScheduleTuesday) geändert. Zeitraum am Vormittag verschoben, Zeitraum am Nachmittag hinzugefügt. Dann passiert nach dem Speichern des Datenpunktes (mehr muss man nicht machen) in der ViCare-App folgendes:

https://github.com/open3e/open3e/assets/144950531/82f86ca1-1fbc-4cf2-84cb-0980f73e8509

**Comment by MyHomeMyData** [comment]
# Anleitung zur Installation

* Der Adapter kann über das Beta-Repository von ioBroker installiert und aktualisiert werden. Normalerweise ist das Stable-Repository ausgewählt, dann muss einmalig auf das Beta-Repo umgestellt werden:
   * Dazu klickt man auf den Schraubenschlüssel für die Basiseinstellungen (1)
   * In den Einstellungen den Tab "Repositories" auswählen (2) und das Häkchen auf "beta" ändern (3). Nach dem Klick auf "Speichern und Schließen" sollte der Adapter auftauchen (5). Sollte das nicht passieren, mit (4) aktualisieren.

* **Bitte bei der ersten Konfiguration ggf. open3e stoppen**, falls das läuft. Sonst könnte es bei den Scans Konflikte auf dem Bus geben.


![Ashampoo_Snap_2024 01 30_18h16m9s_001_adapters - raspi � Mozilla Firefox](https://github.com/open3e/open3e/assets/144950531/30200a0f-ee9b-49ad-bd27-98e5d0f1d693)


![Ashampoo_Snap_2024 01 30_18h17m22s_002_adapters - raspi � Mozilla Firefox](https://github.com/open3e/open3e/assets/144950531/1d4cc6bd-305f-45f0-a108-17cc6c035636)


![Ashampoo_Snap_2024 01 30_18h18m53s_003_adapters - raspi � Mozilla Firefox](https://github.com/open3e/open3e/assets/144950531/4f5a4544-2501-4848-898a-daf145255625)

**Comment by jwende** [comment]
Hut ab … da war aber jemand extrem fleißig. Muss auf dem Raspi laufen - oder?
J.
Gesundes neues Jahr!

**Reply by MyHomeMyData** [reply]
Hallo Jörg, auch Dir ein gutes Neues!

Stimmt, da ist in den letzten Wochen sehr viel Zeit rein geflossen. Bei dem ständigen Regenwetter aber eine willkommene Beschäftigung.

Der Adapter läuft selbst auf einem Raspi Zero, aber zäh. Empfehlen würde ich Raspi 3 oder besser. Prinzipiell sollte der Adapter mit jeder ioBroker-Installation laufen. Wie man aber unter Windows an den CAN-Adapter ran kommt, weiß ich nicht. Habe ich auch noch nicht ausprobiert. Deshalb im Moment vermutlich auf Linux-Unterbau beschränkt.

**Comment by wotandamm** [comment]
Hallo,
der Adapter läuft sehr gut auf einem Raspi3B.
Ich betreibe nur eine VX3 und einen E380. Die Objekte für den E380 werden ja automatisch beim anklicken erzeugt.
Die Objekte für die VX3 werden zusammen mit vielen anderen die ich nicht benötige erzeugt.
Wir kann ich Objekte nur für die VX3 erzeugen? 
Eventuell auch nur bestimmte der VX3?

**Reply by MyHomeMyData** [reply]
Die umfangreichen Objektbäume werden beim Datenpunkt-Scan angelegt. Da werden ja sämtliche Datenpunkte angefordert und abgespeichert.

Wenn Du nur die Objekte in den Objektbäumen sehen möchtest, die Du per UDSonCAN anforderst und die per "Collect" eingesammelt werden, kannst Du folgendes machen:
* Adapter stoppen
* Bei den gewünschten Geräten die Objektbäume "json", "raw" und "tree" komplett löschen, also auch alle Unterstrukturen (Button "Ganzen Objektbaum (xxxx Objekte)").
* Bitte darauf achten, den **Objektbaum "info" des Geräts nicht zu löschen**.
* Adapter wieder starten

Jetzt werden alle Objekte wieder angelegt, die von den Geräten gelesen werden.

In der Adapter-Konfiguration unter "Liste der Datenpunkte" kannst Du weiterhin sämtliche Datenpunkte eines Geräts anzeigen.

**Comment by wotandamm** [comment]
Hallo,

danke, so gefällt es mir schon besser.
Wie kann ich einen Datenpunkt verändern? Hier speziell e3oncan.0.VX3.tree.1684_AmbientTemperatureSensor?
Der Datenpunkt e3oncan.0.VX3.tree.2240_BatteryTemperatureSensor wird korrekt mit den drei Temperaturen angezeigt.
Der Datenpunkt e3oncan.0.VX3.tree.1684_AmbientTemperatureSensor aber nur als HEX Wert: d100c8001201df0000 ?

Die Datenpunkte unter e3oncan.0.e380.tree.0600_GridEnergy werden nicht als Zählerstände angezeigt.
Gibt es mittlerweile eine Formel mit der die Werte in tatsächliche Zählerstände umgerechnet werden können?

Meine Werte:
e3oncan.0.e380.tree.0600_GridEnergy.ExportCumulated = 622150.3945  Zähler = 2793 kWh
e3oncan.0.e380.tree.0600_GridEnergy.ImportCumulated = 619810.5545  Zähler = 1860 kWh

**Reply by MyHomeMyData** [reply]
> Wie kann ich einen Datenpunkt verändern? Hier speziell e3oncan.0.VX3.tree.1684_AmbientTemperatureSensor?

Die Datenpunktliste wird per Skript aus dem open3e-Projekt in ein JSON umgewandelt. Daraus baut der Adapter beim Datenpunktscan die gerätespezifischen Datenpunktlisten auf. Der beste Weg wäre also dieser:
* Neue Erkenntnisse zu Datenpunkten sollten in open3e in die Datei `Open3Edatapoints.py` eingepflegt werden.
* Mit dem nächsten Update der Liste im Adapter fließen diese Infos dann auch im Adapter ein.

In der Zwischenzeit kannst Du das temporär auch direkt in Deinem Objektbaum ändern:
* Adapter stoppen
* Für die VX3 das Objekt `e3oncan.0.VX3.info.udsDidsCommon` bearbeiten und die Definition von 1684 anpassen bzw. aus 2240 kopieren. Leider ist das JSON-Format nicht grade übersichtlich. Falls das Format nach der Änderung nicht mehr passt, könnte der Adapter abstürzen. Also am besten vor der Änderung das Objekt sichern.
* Adapter wieder starten.

Bei einem Update des Adapter wird dieses Objekt überschrieben, falls es eine neuere Version der Datenpunktliste enthält. Änderungen an diesem Objekt sind also nicht dauerhaft.

Falls Du das ausprobierst, gib mir bitte eine Rückmeldung. Habe das selbst noch nicht probiert.

> Die Datenpunkte unter e3oncan.0.e380.tree.0600_GridEnergy werden nicht als Zählerstände angezeigt.

Die Werte sollten bereits auf kWh umgerechnet sein. Allerdings gibt es für jedes Gerät einen unbekannten Offset. Die individuelle Umrechnung sollte nur die Subtraktion eines festen Offsets erfordern. Ist das bei Dir nicht der Fall?

**Comment by wotandamm** [comment]
So, den Wert für 1684 habe ich in Open3E geändert (lassen) 😊.

Den Wert in e3oncan.0.VX3.info.udsDidsCommon kann ich nicht ändern, der Datenpunkt ist bis auf {} leer.

 



 

Den Offset zur Umrechnung vom e380 kenne ich nicht. Lässt der sich aus den Objektwerten und dem abgelesenen Zählerstand bilden?

**Reply by MyHomeMyData** [reply]
> Den Wert in e3oncan.0.VX3.info.udsDidsCommon kann ich nicht ändern, der Datenpunkt ist bis auf {} leer.

Dann hast Du vermutlich keinen Datenpunkt-Scan durchgeführt. Das kannst Du nachholen im Konfigurations-Tab "Liste der Datenpunkte".

Oder Du wartest noch ein paar Tage. Ich arbeite grade an einem Update, da ist der Datenpunkt dann drin.

> Den Offset zur Umrechnung vom e380 kenne ich nicht. Lässt der sich aus den Objektwerten und dem abgelesenen Zählerstand bilden?

Bisher gehe ich davon aus, dass die Offsets zufällig sind. Welchen Zählerstand meinst Du? Wenn Du einen Zählerstand in kWh hast, sollte sich der Offset ja einfach durch Subtraktion bilden lassen, z.B.: Offset = 600_GridEnergy.Import - Zählerstand.
Anschließend müsste 600_GridEnergy-Offset mit dem Zählerstand übereinstimmen. Falls das mit der Zeit deutlich auseinander läuft, hat Dein E380 einen anderen Umrechnungsfaktor.
